File: backend/app/api/v1/endpoints/incidents.py
import uuid

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import CurrentUser, DbSession
from app.models.enums import IncidentStatus
from app.schemas.incident import (
    IncidentAnalyzeRequest,
    IncidentDetail,
    IncidentRead,
    IncidentStatusUpdate,
    PaginatedIncidents,
)
from app.services.incident_service import IncidentNotFoundError, IncidentService

router = APIRouter(prefix="/incidents", tags=["incidents"])


@router.post("/analyze", response_model=IncidentDetail, status_code=status.HTTP_201_CREATED)
async def analyze_incident(
    request: IncidentAnalyzeRequest,
    db: DbSession,
    current_user: CurrentUser,
) -> IncidentDetail:
    incident = await IncidentService(db).analyze_and_create(request, current_user.id)
    return IncidentDetail.model_validate(incident)


@router.get("", response_model=PaginatedIncidents)
async def list_incidents(
    db: DbSession,
    _: CurrentUser,
    status_filter: IncidentStatus | None = Query(default=None, alias="status"),
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
) -> PaginatedIncidents:
    total, incidents = await IncidentService(db).list(
        status=status_filter, limit=limit, offset=offset
    )
    return PaginatedIncidents(
        total=total, items=[IncidentRead.model_validate(i) for i in incidents]
    )


@router.get("/{incident_id}", response_model=IncidentDetail)
async def get_incident(
    incident_id: uuid.UUID,
    db: DbSession,
    _: CurrentUser,
) -> IncidentDetail:
    try:
        incident = await IncidentService(db).get(incident_id)
    except IncidentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Incident not found"
        ) from exc
    return IncidentDetail.model_validate(incident)


@router.patch("/{incident_id}/status", response_model=IncidentRead)
async def update_incident_status(
    incident_id: uuid.UUID,
    update: IncidentStatusUpdate,
    db: DbSession,
    _: CurrentUser,
) -> IncidentRead:
    try:
        incident = await IncidentService(db).update_status(incident_id, update.status)
    except IncidentNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Incident not found"
        ) from exc
    return IncidentRead.model_validate(incident)

File: frontend/src/api/incidents.ts
import type {
  IncidentDetail,
  IncidentStatus,
  LogEntryInput,
  PaginatedIncidents,
} from '@/types';

import { apiClient } from './client';

export interface ListIncidentsParams {
  status?: IncidentStatus;
  limit?: number;
  offset?: number;
}

export async function listIncidents(
  params: ListIncidentsParams = {},
): Promise<PaginatedIncidents> {
  const { data } = await apiClient.get<PaginatedIncidents>('/incidents', { params });
  return data;
}

export async function getIncident(id: string): Promise<IncidentDetail> {
  const { data } = await apiClient.get<IncidentDetail>(`/incidents/${id}`);
  return data;
}

export async function analyzeIncident(
  logs: LogEntryInput[],
  title?: string,
): Promise<IncidentDetail> {
  const { data } = await apiClient.post<IncidentDetail>('/incidents/analyze', {
    title: title ?? null,
    logs,
  });
  return data;
}

export async function updateIncidentStatus(
  id: string,
  status: IncidentStatus,
): Promise<void> {
  await apiClient.patch(`/incidents/${id}/status`, { status });
}
